tools/gema/doc/gema.1
.\" @(#)gema.1
.\" $Id: gema.1,v 1.6 2003/12/28 02:18:03 dngray Exp $
.TH GEMA 1 "26 Dec 2003" "version 1.4" " "
.SH NAME
gema \- general purpose macro processor
.SH SYNOPSIS
\fBgema\fP \fIoptions\fP... [ \fIinput_file\fP [ \fIoutput_file\fP ]]
.LP
\fBgema\fP \fIoptions\fP... \fB\-out\fP \fIoutput_file\fP \fIinput_file\fP...
.LP
\fBgema\fP \fIoptions\fP... \fB\-odir\fP \fIoutput_directory\fP
[\fB\-otyp\fP \fIoutput_suffix\fP]
\fIinput_file\fP...
.LP
Where \fIoptions\fP is:
.br
\fB\-f\fP \fIpattern_file\fP | [\fB\-p\fP] '\fIpatterns\fP' |
.BR \-b " | " \-i " | " \-k " | "
.BR \-line " | " \-match " | "
.BR \-t " | " \-w " | "
\fB\-idchars \fP\fIcharset\fP |
\fB\-filechars \fP\fIcharset\fP |
\fB\-literal \fP\fIcharset\fP |
.BR \-ml " | "
.BR \-trace " | "
\fB\-arglen \fP\fInum\fP
.de QX
"\\fB\\$1\\fP"\\$2
..
.SH DESCRIPTION
.B gema
is a general purpose text processing utility based on the concept of
pattern matching.  In general, it reads an input file and copies it to
an output file, while performing certain transformations to the data as
specified by a set of patterns defined by the user.  It can be used to
do the sorts of things that are done by Unix utilities such as
.BR cpp ,
.BR grep ,
.BR sed ,
.BR awk ,
or
.BR strings .
It can be used as a macro processor, but it is much more general than
.BR cpp " or " m4
because it does not impose any particular syntax for what a macro call
looks like.  Unlike utilities like
.BR sed " or " awk ,
.B gema
can deal with patterns that span multiple lines and with nested constructs.
It is also distinguished by being able to use multiple sets of rules to
be used in different contexts.
.LP
This
.B man
page provides a tutorial introduction and a brief reference summary.
See the user manual for a more detailed specification.
.LP
If no file names are given on the command line, the program reads from
standard input and writes to standard output.  With two file names
provided, it reads from the first and writes to the second.  If the
output file previously existed, the old file is saved with a
.QX .bak
suffix.  The input file and output file may be the same, in which case
the program actually reads from the backup file.
With the
.B \-out
option, it will write to the designated file, concatenating the results
of reading from any number of input files, which may be specified by wild
cards.  With the
.B \-odir
option, it reads any number of input files, creating a corresponding
output file for each in the designated output directory, with the output
file suffix optionally specified by the
.B \-otyp
option.
.LP
.ne 6
.SH "TUTORIAL EXAMPLES"
For a simple example, consider first the following command line:
.br
\ \ \ \fBgema 'Abram=Abraham;Sarai=Sarah' old.text new.text\fP
.br
This copies file "old.text" to "new.text", replacing each occurrence of
the string "Abram" with the string "Abraham" and replacing "Sarai" with
"Sarah".  The first command line argument consists of two transformation
rules separated by '\fB;\fP'.  Each rule consists of a
.I template
followed by '\fB=\fP' and an
.IR action .
Any text that matches a template is replaced by the result of the
corresponding action.  (Here the action is just a literal replacement
string, but it is called an
.I action
instead of a
.I replacement
because it is more general than that.)
The presence of the equal sign serves to distinguish a pattern argument
from a file name.  Note that pattern arguments will usually need to be
enclosed in single quotes on Unix (as shown in the example), or double
quotes on MS-DOS or Windows.
.LP
Patterns can also be defined in one or more files loaded with the
.B \-f
option.  In a pattern file, new lines separate rules, blank lines are
ignored, and a '\fB!\fP' causes the rest of the line to be ignored as a
comment.
The following characters have special meaning in patterns:
.br
\ \ \ \fB: * ? # / < > \\ ^ = $ @ { } ; !\fP Space NewLine
.LP
Usually when doing string replacement we need to be a little more careful.
For example, suppose we want to replace the variable name "x" with the
name "horizontal".  We don't want to replace instances of "x" that appear
as parts of words, only those that are a word by themselves.
This can easily be done like this:
.br
\ \ \ \fBgema \-t 'x=horizontal'\fP ...
.br
where the
.B \-t
option (token mode) requires all identifiers in templates to match a
complete identifier in the input data.
If you don't want this behavior everywhere, it can instead be specified
locally in the template like this:
.br
\ \ \ \fBgema '\\Ix\\I=horizontal'\fP ...
.br
Here the notation "\fB\\I\fP" denotes an identifier delimiter; it
matches an empty string but only if one of the adjacent characters can
not be part of an identifier.  There are a number of other such
operators listed in the reference section below.  Another that is
particularly useful is "\fB\\W\fP", which says that any number of
whitespace characters will be accepted and skipped over if they appear
at that point.  Thus a template like "\fBx\\W+\\Wy\fP" will match regardless
of whether the input text contains optional spaces.
The
.B \-w
option may be used to ignore spaces everywhere (except within identifiers
or numbers).
.LP
A template such as "\fBfirst\ down\fP" will not only match these two
words separated by a space, it will also match if they are separated by
multiple spaces or even by a newline.  This is because the space
character actually has special meaning, which is to match one or more
whitespace characters.  In the less likely event that you really want to
match exactly one space character, you can use "\fB\\\ \fP" or "\fB\\s\fP".
.LP
Templates can also have
.I arguments
-- i.e. portions which match variable text.  There are in fact several
different kinds of arguments supported.  The first kind of argument is
the
.I "wild card"
characters "\fB*\fP", which matches any number of characters (up to a
limit that can be adjusted by the
.B \-arglen
option), and
"\fB?\fP", which matches any one character.  For example, consider the C
macro definition:
.br
\ \ \ \fB#define frob(p,m) do(Frob,p[m])\fP
.br
A first approximation to doing this with
.B gema
is to use a rule like this:
.QX \\\\Ifrob\\\\W(*,*)=do(Frob,*[*])
Here the asterisks in the template match any text up to the delimiting
comma or
parenthesis, and the same text is substituted where the asterisks appear
in the action.  Note though that in
.BR gema ,
the comma and parentheses do not have any special meaning; a template
may delimit the arguments with whatever characters you want.
There may be as many as 20 arguments, and the "\fB*\fP" or "\fB?\fP"
in the action will be substituted in the same order as in the template.
In cases where the arguments need to be used in some different order or
an argument needs to be used more than once, a different notation may be
used that designates the arguments by number.  For example:
.br
\ \ \ Rule:  \fBADD * TO *.=$2 \\:\\= $2 + $1\\;\fP
.br
\ \ \ Input: \fBADD ITEM TO SUM.\fP
.br
\ \ \ Output: \fBSUM := SUM + ITEM;\fP
.br
Note that some of the characters are escaped so that they will be
treated as literals.  A backslash preceding a special character always
means that the character is to be taken literally, so it won't hurt to
escape all special character literals if you aren't sure which ones have
special meaning.
(If there are more than nine arguments, the number needs to be enclosed in
braces, like: "\fB${10}\fP")
.LP
One problem to watch out for with "\fB*\fP" arguments is that since
.B gema
allows patterns to span line boundaries, the argument may read many
lines ahead, gobbling up much more text than intended.  In cases where
you want to match only on a single line, the operator "\fB\\L\fP" can be
used in a template to specify that the following arguments will not be
allowed to match on a newline character.  This mode may also be selected
globally with the command line option
.BR \-line .
.LP
When a template is to match an entire line, the first inclination might be to
write a template such as:
.QX "\\\\nCommand *\\\\n"
where
.QX \\\\n
designates the new line character.
But there are two problems with this:  it won't match on the first line
of the file (because there isn't a preceding new line) and it won't
match on lines where the preceding new line has already been read as
part of the match for the previous line.  Both of these are solved by
using
.QX \\\\N
which matches if it is at the beginning or end of a line, but doesn't
actually read the new line character.  In an action,
.QX \\\\N
outputs a new line character if the output stream is not already at the
beginning of a line.
.LP
Another way to further constrain argument matching is to use the next
kind of argument, which is called a
.IR "predefined recognizer domain" .
(We'll see user-defined domains later.)  For example, in a template the
string "\fB<D>\fP" designates an argument that will match one or more
digits.  Similarly,
"\fB<L>\fP" matches one or more letters,
"\fB<I>\fP" matches an identifier (letters, digits, and underscore),
and "\fB<N>\fP" matches a number (digits with optional sign and decimal
point).  Other kinds are listed in the reference section below.
With a lower case letter, the argument becomes optional.  Thus,
"\fB<d>\fP" matches zero or more digits.  Preceding the letter with a
minus sign inverts the test, so that, for example, "\fB<\-D>\fP" matches
one or more characters that are
.I not
digits.  The letter may be followed by a number to limit how many
characters are matched.  For example, "\fB<D3>\fP" matches exactly three
digits and "\fB<d3>\fP" will match from zero up to a maximum of three digits.
A
.I universal
argument such as "\fB<U10>\fP" matches exactly 10 characters of any
kind; this may be useful for splitting an input record into fixed-length
fields.  Note that "\fB<U1>\fP" is equivalent to "\fB?\fP".
(For these arguments, the action must access the value using
"\fB$\fP" instead of repeating the argument designator.  That short cut
is only supported for arguments designated by a single character.)
.LP
An even more general way of specifying the set of characters to be
accepted is to use a
.IR "regular expression" .
A regular expression argument is designated by slashes before and after.
For example, the template "\fBc/[ad]+/r\fP" will match input text "cadar"
with "\fB$1\fP" having the value "ada".  Note that between the slashes,
the special characters have the meaning defined by the usual regular
expression syntax, not their meaning elsewhere in templates.
.LP
The last kind of argument is a
.IR "recursive argument" .
To show the need for this, first consider this example that is
trying to convert Lisp s-expressions to function call notation:
.br
\ \ \ Rule: \fB(* * *)=*(*,*)\fP
.br
\ \ \ Input: \fB(fn xyz 34)\fP
.br
\ \ \ Output: \fBfn(xyz,34)\fP
.br
So far, so good, but look what happens here:
.br
\ \ \ Input: \fB(fn (g a b) z)\fP
.br
\ \ \ Output: \fBfn((g,a b)\fP
.br
What is needed is a way to properly associate matching nested
parentheses and to translate nested constructs.  Both of these are done
by simply modifying the rule like this:
.br
\ \ \ Rule: \fB(# # #)=#(#,#)\fP
.br
\ \ \ Input: \fB(fn (g a b) z)\fP
.br
\ \ \ Output: \fBfn(g(a,b),z)\fP
.br
The
.QX #
designates a
.IR "recursive argument" ,
which means that the argument value is
the result of translating input text until reaching the terminator
character(s) following it.  The space following "g" does not terminate
the outer-level argument because it is read as part of the inner-level
template match.  Similarly, the inner "(" is read by the inner match
which also reads the matching inner ")".
.LP
Actions can also perform a wide variety of activities by using the large
set of built-in functions that are available.  A function call is
designated by
.QX @
followed by the name of the function, followed by any arguments enclosed
in curly braces and separated by semicolons.  The "functions" section
below lists all of the functions that are available.
For example, you can define a default rule like this:
.QX "\\\\N.*\\\\N=@err{@file line @line\\\\: Unrecognized\\\\: $1}"
The
.B err
function writes its argument to the error output stream.  The
.B file
and
.B line
functions (which have no arguments) return the name and line number of
the input file.
.LP
.B gema
supports defining multiple sets of rules, each of which is called a
.IR domain.
By default, rules are defined in the default domain, which is also the
one used to translate the input file.  Alternate domains are created by
prefixing a rule with a domain name followed by
.QX : .
The domain name applies until the end of the line.
The name of the default domain is the empty string, so a rule line
beginning with a colon is the same as one without any colon.
Alternate domains are used for several purposes, including
defining new kinds of arguments for use in templates, defining new
functions for use in actions, and for translations that require
different rules for various contexts in the input data.
.LP
To illustrate using an alternate domain as a new argument type, suppose
we want a template argument that will match on either "yes" or "no", so
that we can write a rule like:
.br
.B "   done\\\\? <yesno>=Finished \\\\= $1"
.br
All that is needed is to define the following:
.br
.B "   yesno:yes=yes@end;no=no@end;=@fail"
.br
This says that if you see "yes" or "no", output it to the argument value and
then return, and if anything is seen that doesn't match either of those,
then the argument match fails.
Note that the last rule has an empty
template; this matches as a last resort if no other template in the
domain can be matched.  Since it doesn't advance the input stream, this
makes sense only if the action is to exit.
Note that domain names should have at least two letters in order to not
conflict with predefined recognizers.
.LP
Domain names can also be used as functions of one argument, which means
that the function returns the result of translating the argument value
with the patterns of the domain.  This is typically used in a two-step
translation process where the first pattern match is used to split the
input into fields, and then other domains are used to process individual
fields.  Remember that the default domain has an empty name, so
"\fB@{\fP\fIarg\fP\fB}\fP" means to translate the argument with the
default domain.
.LP
For an example of the use of alternate domains for different contexts,
suppose that we are doing name substitution in C source code and we
don't want to make
any changes inside of character strings.  We could add a rule
.QX """*""=""*"""
to match on string constants and pass them through.  However,
string constants can contain
.QX "\\\\"""
and we don't want the argument to be terminated by that quote.  To handle
this, we can use a separate domain for processing the contents of a string.
Then the rule becomes
.QX """<sbody>""=""$1"""
and we add a rule:
.QX "sbody:\\\\\\\\""=\\\\\\\\"""
.LP
.ne 6
.SH OPTIONS
.LP
.TP
.BI \-f " pattern_file"
Reads pattern definitions from a file.
.TP
.BI \-p " patterns"
Patterns defined directly as a command line argument.  The
.B \-p
can usually be omitted since an argument containing
.QX =
or beginning with
.QX @
is automatically recognized as a pattern argument.
Any number of 
.B \-f
and/or
.B \-p
options may be given.
.TP
.BI \-in " file"
Explicitly specifies the input file pathname.  If the file name is
.QX -
then standard input is used.  Usually the
.B \-in
is not necessary since the first file name on the command line is
understood to be the input file.
.TP
.BI \-out " file"
Specifies the pathname of the output file.  If the name is
.QX -
then standard output is used.  After an explicit
.B \-out
option has been used, the remainder of the command line can have any
number of input file names (without
.BR \-in )
which will be read in sequence, with the
concatenated result going to the single output file.
For example, a command such as the following can be used to do a 
.BR grep -like
search of a group of files:
.br
\ \ \fBgema -match -p 'Copyright *\\n=@file\\: $0' -out - *.c\fP
.br
(The special notation
.QX $0
copies all of the matched text into the output.)
.TP
.BI \-odir " directory"
Specifies the output directory.  For each input file that follows, a
corresponding output file will be written in the designated directory.
.TP
.BI \-otyp " suffix"
When used with the
.B \-odir
option, this specifies that each output file will have the designated
suffix replacing the suffix of the input file.
For example, given the command line:
.br
\ \ \ \fBgema \-f patterns \-odir /stuff \-otyp .list *.text\fP
.br
then if the current directory contains a file named
.QX foo.text ,
it will be translated to an output file named
.QX /stuff/foo.list .
.TP
.BI \-backup " suffix"
The argument will be used as the backup file suffix in place of the default
.QX .bak .
.TP
.BI \-nobackup
Output files will be overwritten instead of saving the old file as a
backup file.
.TP
.B \-line
Places the program in line mode, which means that all pattern matching
is limited to single lines.  Arguments and template operators will never
cross a line boundary except where the template explicitly specifies
so with
.QX \\\\n .
.TP
.B \-b
Binary.  With this option, all input and output files are opened in
binary mode instead of text mode.  This makes no difference on most Unix
systems, but on Windows it changes the meaning of the new line character and
doesn't treat Control-Z as the end of the file.
.TP
.B \-k
Keep going.  With this option, the program will try to continue
execution after certain errors that would normally cause it to abort.
This may be useful when you want to see everything that is wrong before
starting to fix the errors.  Errors will still cause a non-zero exit
status despite this option.
.TP
.B \-match
Matches only mode.  Input text that doesn't match any template will be
discarded instead of being copied to the output.  This would be used
when you want to extract selected information (like with
.BR grep )
instead of doing a translation.
This option applies only to the default domain.
Another way to discard otherwise unmatched text is with the default rule
.QX ?=
while the rule
.QX ?=?
can be used to explicitly copy.
.TP
.B \-i
Case insensitive mode.  All letters in templates will be matched without
regard to distinctions of upper case or lower case.  This also makes the
names of domains and built-in functions case insensitive.
.TP
.B \-w
Whitespace insensitive mode.  Space and tab characters in rules will be
ignored except where they separate identifiers.
Template matching will ignore whitespace characters in the
input data as though templates had an implicit
.QX \\\\W
everywhere except within identifiers.
Templates can use
.QX \\\\J
to indicate where space is
.I not
allowed.
.TP
.B \-t
Token mode.  All identifiers appearing in templates will match only
against a complete identifier, as though each identifier was implicitly
surrounded by
.QX \\\\I
except where counter-acted by
.QX \\\\J .
.TP
.BI \-idchars " charset"
Identifier characters.  The argument value specifies the set of
characters that will be considered to be identifier constituents, in
addition to letters and digits.  The default value is
.QX _ .
This affects the behavior of
.QX \\\\I ,
.QX <I> ,
and
.QX <Y> ,
and the
.B \-w
and
.B \-t
options.
For example, if you were processing
.SM COBOL
source code, you would need
.QX "\-idchars '-'".
For Lisp code, you would probably want something like:
.QX "\-idchars '-+=*/_<>'"
.TP
.BI \-filechars " charset"
File name characters.  The argument value specifies the set of
characters which are accepted by
.QX <F>
as being file name constituents, in addition to letters and digits.
The default value is
.QX "./\-_~#@%+="
for Unix.  On Windows, colon and backslash are also included in the
default set.
.TP
.BI \-literal " charset"
This option specifies that each of the characters in the argument value
will be treated as an ordinary literal character in patterns, instead of
whatever special meaning it might normally have.
For example, rather than saying something like:
.br
 \ \ \ \fBgema '\\/usr\\/foo\\/<F>=\\/usr\\/bar\\/$1'\fP \fI...\fP
.br
you could instead say:
.br
 \ \ \ \fBgema -literal / '/usr/foo/<F>=/usr/bar/$1'\fP \fI...\fP
.TP
.B \-ml
For convenience in processing Markup Languages (HTML, XML, etc.), this
option (which is new in version 1.4) 
changes the syntax so that the characters
.QX < ,
.QX >
and
.QX /
are taken as literals, using 
.QX [ ,
.QX ]
and
.QX |
respectively in their places.
For example:
.br
 \ \ \ \fBgema -ml -p '<i>[T]</i>=<em>$1</em>'\fP \fI...\fP
.br
has the same effect as:
.br
 \ \ \ \fBgema -p '\\<i\\><T>\\<\\/i\\>=\\<em\\>$1\\<\\/em\\>'\fP \fI...\fP
.br
The
.QX -ml
option is an abbreviation for
.QX @set-syntax{</>LLL;[|]</>} .
.TP
.BI \-arglen " number"
Specify the maximum length of a
.QX *
argument.  The default is 4096.
.TP
.BI \-prim " pattern_file"
Primitive mode (for advanced users only).
Like the
.B \-f
option, this loads patterns from a file.  It also suppresses loading of
the built-in patterns for command line processing.  This option is
meaningful only when it appears as the first argument, and then it
becomes the only argument that has any predefined meaning.  The
designated pattern file must define
.B ARGV
domain rules sufficient to specify what to do with the remainder of the
command line.
.TP
.B \-help
Display brief usage message on the standard error output.
.TP
.B \-version
Display program version identification on the standard error output.
.TP
.B \-trace
If the program was compiled with
.BR \-DTRACE,
then this option can be used to enable a report of template matches
and failures to be written to the error output.  This may be helpful for
diagnosing obscure cases of unexpected results, but this is a crude
experimental feature, so don't expect too much.
Each line describes one event, possibly showing the line and column number
at the beginning of the current template, the line and column of the
current position, and a description of the event, indented for recursive
arguments. 
.RE
.LP
.ne 6
.SH "PATTERNS REFERENCE SUMMARY"
The following characters have special meaning:
.LP
.TP
.B *
matches any number of characters
.TP
.B ?
matches any one character
.TP
.B #
argument recursively translated in the current domain
.TP
.B =
end of template, beginning of action
.TP
.B $0
copies the template into the action to show all matched text
.TP
.BI $ digit
inserts argument value
.TP
.BI $ letter
inserts value of a variable with single-letter name
.TP
.BI ${ name }
value of named variable (only in action)
.TP
.BI ${ name ; default }
variable with default value if not defined (action only)
.TP
\fB\e\fP
escape character; see the section on "escape sequences" below.
.TP
.BI ^ x
combine control key with the following character
.TP
.I Space
matches one or more whitespace characters (same as
.QX \\\\S ).
.TP
.I NewLine
end of action
.TP
.B ;
end of action, or separator between function arguments
.TP
.BI @ name { args }
invoke built-in function or user-defined translation domain (action only)
.TP
.BI @ special-character
has the default meaning of the special character by itself, as
documented here; this can be used to access the original functionality of a
character that has been changed by the
.B \-literal
option or
.B @set-syntax
function.
.TP
.B :
separates domain name from rule
.TP
.BI < name >
recursive argument, translated according to the named domain,
or pre-defined recognizer argument.  (template only)

tools/gema/src/gema.c

/* generalized macro processor */

/* $Id: gema.c,v 1.14 2001/12/15 20:22:13 gray Exp $ */

/*********************************************************************
  This file is part of "gema", the general-purpose macro translator,
  written by David N. Gray <dgray@acm.org> in 1994 and 1995.
  You may do whatever you like with this, so long as you retain
  an acknowledgment of the original source.
 *********************************************************************/

/*
 * $Log: gema.c,v $
 * Revision 1.14  2001/12/15 20:22:13  gray
 * Add new convenience option "-ml" for using "[...]" instead of "<...>"
 * with HTML, XML, etc.
 *
 * Revision 1.13  1995/10/01  23:29:44  gray
 * Fix to support MS-DOS wild card expansion with a full pathname.
 *
 * Revision 1.12  1995/09/29  05:42:18  gray
 * Fix MS-DOS version for input file specified as full pathname.
 *
 * Revision 1.11  1995/08/07  03:21:25  gray
 * Remove support for "/" options in MS-DOS because doesn't work right.
 *
 * Revision 1.10  1995/07/04  23:41:52  gray
 * Split `argv_rules' into 5 pieces to placate the MPW compiler.
 *
 * Revision 1.9 1995/06/12 02:59:42 gray
 * Add "-trace" option.  Fix "-f" to not undo binary mode set in file.
 *
 * Revision 1.8 1995/05/22 02:50:14 gray
 * Update expand_wildcard to work on Windows/NT.
 *
 * Revision 1.7 1995/05/08 03:15:15 gray
 * Add wild card expansion for MS-DOS
 */

#if defined(_QC) || defined(_MSC_VER)
#pragma check_stack(off)
#endif

#include <stdio.h>
#include <ctype.h>	/* for isspace */
#include <string.h>
#include <assert.h>
#include "cstream.h"
#include "pattern.h"
#include "util.h"
#include "main.h"

COStream output_stream;
COStream stdout_stream;
CIStream stdin_stream;

boolean keep_going = FALSE;
boolean binary = FALSE;
Exit_States exit_status = EXS_OK;

void usage(void) {
  fprintf(stderr,
	  "Syntax: gema [ -p <patterns> ] [ -f <pattern-file> ] [ <in-file> <out-file> ]\n");
  fprintf(stderr, "Copies standard input to standard output, performing"
	  "\ntext substitutions specified by a series of patterns\n"
	  "specified in -p options or in files specified by "
	  "-f options.\n");
  pattern_help( stderr );
}

static struct switches {
  const char* name;
  boolean * var;
} switch_table[] =
  { { "line", &line_mode },
    { "b", &binary },
    { "k", &keep_going },
    { "match", &discard_unmatched },
    { "i", &case_insensitive },
    { "w", &ignore_whitespace },
    { "t", &token_mode },
    { "arglen", &MAX_ARG_LEN },
#ifndef NDEBUG
    { "debug", &debug_switch },
#endif
#ifdef TRACE
    { "trace", &trace_switch },
#endif
    { NULL, NULL } };

int*
find_switch(const char* arg) {
  /* given a switch name, return a pointer to the value. */
  struct switches *p;
  for ( p = &switch_table[0] ; p->name != NULL ; p++ )
    if ( stricmp(arg, p->name)==0 )
      return p->var;
  return NULL;
}

static struct parms {
  const char* name;
  char ** var;
} parm_table[] =
  { { "idchars", &idchars },
    { "filechars", &filechars },
    { "backup", &backup_suffix },
    { NULL, NULL } };

boolean
set_parm(const char* name, const char* value) {
  struct parms *p;
  for ( p = &parm_table[0] ; p->name != NULL ; p++ )
    if ( stricmp(name, p->name)==0 ) {
      char** vp = p->var;
#if 0 /* don't bother with this to avoid copying the initial value. */
      if ( *vp != NULL )
	free(*vp);
#endif
      *vp = str_dup(value);
      return TRUE;
    }
  return FALSE;
}

#ifdef MSDOS
#ifdef __TURBOC__
#include <dir.h>
#endif
#ifdef _WIN32
#include <io.h>
#else
#include <dos.h>
#endif
#endif /* MSDOS */

void
expand_wildcard ( const char* file_spec, COStream out ) {
/*  Expand wild card file name on MS-DOS.
    (On Unix, this is not needed because expansion is done by the shell.)
 */
#ifdef MSDOS						       
#if defined(_FIND_T_DEFINED)  /* Microsoft C on MS-DOS */
  struct _find_t fblk;
  if ( _dos_findfirst( file_spec, _A_NORMAL|_A_ARCH|_A_RDONLY, &fblk )
       == 0 ) {
    /* first match found */
    do {
      merge_pathnames( out, FALSE, file_spec, fblk.name, NULL );
      cos_putch( out, '\n' );
    }
    while ( _dos_findnext( &fblk ) == 0 );
  }
  else
#elif defined(_WIN32)  /* Microsoft C/C++ on Windows NT */
  struct _finddata_t fblk;
  long handle;
  handle = _findfirst( (char*)file_spec, &fblk );
  if ( handle != -1 ) {
    /* first match found */
    do {
      if ( !(fblk.attrib & _A_SUBDIR) ) {		   
	merge_pathnames( out, FALSE, file_spec, fblk.name, NULL );
	cos_putch( out, '\n' );
      }
    }
    while ( _findnext( handle, &fblk ) == 0 );
    _findclose( handle );
  }
  else
#elif defined(__TURBOC__)  /* Borland Turbo C */
  struct ffblk fblk;
  if ( findfirst( file_spec, &fblk, FA_ARCH|FA_RDONLY ) == 0 ) {
    /* first match found */
    do {
      merge_pathnames( out, FALSE, file_spec, fblk.ff_name, NULL );
      cos_putch( out, '\n' );
    }
    while ( findnext( &fblk ) == 0 );
  }
  else
#endif /* Borland */
  if ( strchr(file_spec,'*') != NULL )
    fprintf( stderr, "No match for \"%s\"\n", file_spec );
  else
#endif /* MSDOS */
  {
    cos_puts( out, file_spec );
    cos_putch( out, '\n' );
  } 
}

static char argv_domain_name[] = "ARGV";

static void
do_args(char** argv) {
  /* process the arguments according to the ARGV pattern domain */
  char** ap;
  CIStream argsbuf;
  COStream outbuf;
  const char* remaining;
  boolean ok;

  outbuf = make_buffer_output_stream();
  cos_putch(outbuf, '\n');
  for ( ap = argv ; *ap != NULL ; ap++ ) {
    cos_puts(outbuf, *ap);
    cos_putch(outbuf, '\n');
  }
  argsbuf = convert_output_to_input( outbuf );
  outbuf = make_buffer_output_stream();
  ok = translate ( argsbuf, get_domain(argv_domain_name), outbuf, NULL );
  cis_close(argsbuf);
  argsbuf = convert_output_to_input( outbuf );
  remaining = cis_whole_string(argsbuf);
  while ( isspace(*remaining) )
    remaining++;
  if ( remaining[0] != '\0' ) {
    fprintf(stderr, "Unrecognized arguments:\n%s", remaining);
    ok = FALSE;
  }
  cis_close(argsbuf);
  if ( !ok )
    exit_status = EXS_ARG;
}

#ifdef MSDOS
/* for MS-DOS conventions, use case-insensitive comparison for options */
#define CI "\\C"
#else
#define CI
#endif

/* The following rules define how the command-line arguments will
   be processed. (They are split into several strings to accommodate the
   Macintosh MPW compiler, which does not allow strings longer than 512.) */
static char argv_rules1[] =
CI "\\N-h*\\n=@show-help@end\n"
   "\\A\\n\\Z=@err{@version\\N@show-help}@end\n"
CI "\\N-version\\n=@err{@version\\N}\n"
CI "\\N-f\\n*\\n=@set{.BINARY;@get-switch{b}}@set-switch{b;0}"
		"@define{@read{*}}@cmpn{${.BINARY;0};0;;;@set-switch{b;1}}\n"
CI "\\N-p\\n*\\n=@define{*}\n"
   "\\N-<L1>\\n=@set-switch{$1;1}\n"
CI "\\N-w\\n=@set-switch{w;1}@set-syntax{S;\\s\\t}\n"
CI "\\N-t\\n=@set-switch{w;1}@set-switch{t;1}\n";
static char argv_rules2[] =
CI "\\N-arglen\\n<D>\\n=@set-switch{arglen;$1}\n"
CI "\\N-idchars\\n*\\n=@set-parm{idchars;$1}\n"
CI "\\N-filechars\\n*\\n=@set-parm{filechars;$1}\n"
CI "\\N-literal\\n*\\n=@set-syntax{L;$1}\n"
#ifndef NDEBUG
CI "\\N-debug\\n=@set-switch{debug;1}\n"
#endif
#ifdef TRACE
CI "\\N-trace\\n=@set-switch{trace;1}\n"
#endif
CI "\\N-line\\n=@set-switch{line;1}\n"
CI "\\N-match\\n=@set-switch{match;1}\n"
CI "\\N-ml\\n=@set-syntax{</>LLL;[|]</>}\n" /* for Markup Language (XML,HTML)*/
#ifdef MSDOS
#if 0 /* doesn't work because causes "\Z" action to be invoked. */
   "\\N\\/<L1>\\n=@ARGV{-$1\\n}\n"	/* allow "/" instead of "-" */
#endif
#else
CI "\\N-n\\n=@set-switch{match;1}\n"	/* like for sed */
CI "\\N-e\\n*\\n=@define{*}\n"		/* like for sed */
#endif
;
static char argv_rules3[] =
CI "\\N-nobackup\\n=@set-parm{backup;}\n"
CI "\\N-backup\\n<G>\\n=@set-parm{backup;$1}\n"
CI "\\N-out\\n*\\n=@set{.OUT;$1}\n"
CI "\\N-in\\n*\\n=@set{.IN;$1}\n"
   "\\N\\L*\\=*\\n=@define{$0}\n"
   "\\N\\L\\@*\\n=@define{$0}\n"
CI "\\N-odir\\n*\\n=@set{.ODIR;*}\n"
CI "\\N-otyp\\n*\\n=@set{.OTYP;*}\n"
   "\\N-*\\n=@err{Unrecognized option:\\ \"-*\"\\n}@exit-status{3}\n"
   "\\n=\n"
#ifdef MSDOS
   /* On MS-DOS, file name wild cards have to be expanded by the
      program, unlike Unix where expansion is done by the shell. */
   "\\N*\\n=@ARGV-F{@expand-wild{*}}\n"
   "ARGV-F:*\\n=@ARGV-FILE{${.ODIR;}\\n${.OUT;}\\n${.IN;}\\n*}\n"
#else
   "\\N*\\n=@ARGV-FILE{${.ODIR;}\\n${.OUT;}\\n${.IN;}\\n*}\n"
#endif
   "\\Z=@ARGV-END{${.OUT;}\\n${.IN;}\\n${.ODIR;}\\n}\n";
static char argv_rules4[] =
"ARGV-FILE:\\n\\n\\n<U>=@set{.IN;$1};"
 "\\n\\n<U>\\n<U>=@set{.OUT;$2};"
 "\\n<U>\\n<U>\\n<U>=@err{More than two files specified.\\n}@exit-status{3};"
 "\\n<U>\\n*\\n<U>=@write{$1;@{@read{$3}}};"
 "<U>\\n\\n*\\n<U>=@bind{.OUT;@makepath{$1;@relpath{$3;$3};${.OTYP;}}}"
  "@write{${.OUT};@{@read{$3}}}@close{${.OUT}}@unbind{.OUT};"
 "<U>\\n<U>\\n=@err{Not meaningful\\:\\ both\\ -out\\ and\\ -odir\\n}"
	"@exit-status{3}@end\n";
static char argv_rules5[] =
"ARGV-END:\\n\\n<U>\\n=@end;" /* -odir was specified */
 "<U>\\n\\n\\n=@end;"
 "<U>\\n<U>\\n=@write{$1;@{@read{$2}}}@end;" /* output and input files */
 "\\n\\n\\n=@write{-;@{@read{-}}};" /* no files specified, use stdin/stdout */
 "\\n<U>\\n=@write{-;@{@read{$1}}}@end\n"
#ifndef NDEBUG
  "ARGV-FILE:*=@err{Internal failure in ARGV-FILE \"*\"\\n}\n"
  "ARGV-END:*=@err{Internal failure in ARGV-END \"*\"\\n}\n"
#endif
;

static void
load_rules( const char* argv_rules, size_t length ) {
  CIStream str;
  assert( length < 512 );
  str = make_string_input_stream(argv_rules, length, FALSE);
  read_patterns ( str, argv_domain_name, FALSE );
  cis_close(str);
}

#define LOAD_RULES(argv_rules) load_rules(argv_rules, sizeof(argv_rules)-1)

static void
initialize_argv_domain(void) {
  assert( (int)EXS_ARG == 3 ); /* to match "@exit-status" above */
  LOAD_RULES(argv_rules1);
  LOAD_RULES(argv_rules2);
  LOAD_RULES(argv_rules3);
  LOAD_RULES(argv_rules4);
  LOAD_RULES(argv_rules5);
}

int main(int argc, char* argv[]){
  char** ap;
  initialize_syntax();
  ap = argv+1;
  stdout_stream = make_file_output_stream(stdout,"");
  stdin_stream = make_file_input_stream(stdin,NULL);
  output_stream = stdout_stream;
  if ( argc >= 3 && stricmp(*ap,"-prim") == 0 ) {
    CIStream ps;
    ap++;
    ps = open_input_file( *ap, FALSE );
    if ( ps != NULL ) {
      read_patterns(ps, "", FALSE);
      cis_close(ps);
    }
    ap++;
  }
  else initialize_argv_domain();
  do_args(ap);
  return (int)exit_status;
}

